### PR TITLE
Introduce benchgc: a benchmark tool for correctness check

### LINKED ISSUE #58: Correctness check for single op
- Cover opset in oneDNN graph.
- Use numpy or pytorch as reference.
- Implement similar data filing as oneDNN.
- Set same numerical tolerance as oneDNN.

### PR DESCRIPTION
**Pull request by WangJialei-A**
Resolve #58 
The single op test is demostrated at [here](https://github.com/intel/graph-compiler/actions/runs/9951109863/job/27490087503?pr=161)

The Doc PR is #179

### CODE REVIEW

File: test/benchgc/CMakeLists.txt
Comment (Menooker) [inline test/benchgc/CMakeLists.txt:30]: Is the installation of benchgc enabled as a default target of cmake? I think it should only be triggered when the user requests to.

Comment (WangJialei-A) [reply test/benchgc/CMakeLists.txt:30]: @Menooker 
I'm not quite sure about this. The configuration `-DGC_BENCH_ENABLE` is default `ON`
It behaves different between `make` and `ninja`. `ninja` will not build `benchgc` as default, but `make` will do.
I'm curious about how to control it in `cmake`

Comment (Menooker) [reply test/benchgc/CMakeLists.txt:30]: If running `ninja` will not build `benchgc` by default, that is what I wanted. We are good here. :)

**Comment by leshikus** [comment]
I wonder if we can make PRs on per-person basis in future. This seems like a mixed work of several developers.

**Comment by WangJialei-A** [comment]
> I wonder if we can make PRs on per-person basis in future. This seems like a mixed work of several developers.
> 
> This is just a generic open source principle when different people do separate contributions

Hi @leshikus 
Since this PR contains dependency of other PR, I merged some commit from other members.
I will purge this PR after those dependencies got merged to follow the principle.

**Comment by leshikus** [comment]
@WangJialei-A this is great; this helps separating work of other contributors

File: .github/workflows/build.yml
Comment (xurui1995) [inline .github/workflows/build.yml:52]: for the workaround here,  the root casue is the mlir upstream did not adding this,  and maybe i can create a fix PR in upstream.

Comment (WangJialei-A) [reply .github/workflows/build.yml:52]: We can remove this after your upstream PR got merged

Comment (WangJialei-A) [reply .github/workflows/build.yml]: `find_namespace_packages` resolve this problem

File: test/benchgc/README.md
Comment (xurui1995) [inline test/benchgc/README.md:102]: could you add a more complex example with the use of `filling` and `cmp`? and could you also add an example of the result output here?

Comment (WangJialei-A) [reply test/benchgc/README.md]: @xurui1995 
I have updated the example in the new version. Please have another review. I am not sure whether it satisfies your requirement.

File: test/benchgc/src/benchgc/mlir/arg.py
Comment (xurui1995) [inline test/benchgc/src/benchgc/mlir/arg.py:139]: you can change `import gc_mlir.dialects.arith` to `from gc_mlir.dialects  import arith`, then just use `arith.ConstantOp`

File: test/benchgc/src/benchgc/mlir/arg.py
Comment (xurui1995) [inline test/benchgc/src/benchgc/mlir/arg.py:138]: can we use `ir.OpView`?

File: test/benchgc/src/benchgc/mlir/module.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/mlir/module.py:50]: @xurui1995 
This function is only for single op test.
Unify the function is meaningful when we need to handle arbitrary length of input and output. 
If we only have two cases `i2o1` and `i1o1`, just keep the function separately and make things clear

File: test/benchgc/src/benchgc/mlir/arg.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/mlir/arg.py:139]: @xurui1995 
We have a package named `arith` inside `benchgc`. I think the full name path will be better to handle the ambiguous problem.

File: test/benchgc/src/benchgc/runner.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/runner.py:31]: @xurui1995 
Nice catch! To resolve this problem elegantly, we need to perform a refactor on how to handle op return value

File: test/benchgc/src/benchgc/mlir/arg.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/mlir/arg.py:138]: @xurui1995  
What's the difference between them?

File: test/benchgc/src/benchgc/mlir/module.py
Comment (xurui1995) [reply test/benchgc/src/benchgc/mlir/module.py:49]: I believe there are some single ops with more than two inputs, you can check some quantization-related ops such as `linalg.quantized_matmul`, I think it would be better to have a general init module method here

File: .github/workflows/style.yml
Comment (WangJialei-A) [reply .github/workflows/style.yml]: resolved

File: test/benchgc/src/benchgc/runner.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/runner.py]: resolved

File: test/benchgc/src/benchgc/__main__.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/__main__.py]: resolved

File: test/benchgc/src/benchgc/runner.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/runner.py]: @xurui1995 
Refactored and resolved

File: test/benchgc/src/benchgc/mlir/module.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/mlir/module.py]: resolved

**Comment by ZhennanQin** [comment]
The llvm revision is updated to include the needed fix from upstream. Please rebase and check the CI result.

File: test/benchgc/src/benchgc/__main__.py
Comment (xurui1995) [inline test/benchgc/src/benchgc/__main__.py:248]: we also have a variable called `entry` in line 211 with different types, maybe we can change the name of one of them.

File: test/benchgc/src/benchgc/util.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/util.py]: resolved

File: test/benchgc/src/benchgc/__main__.py
Comment (WangJialei-A) [reply test/benchgc/src/benchgc/__main__.py]: resolved

File: test/benchgc/src/benchgc/mlir/arg.py
Comment (xurui1995) [reply test/benchgc/src/benchgc/mlir/arg.py:138]: they are the same, `OpView ` is from the `ir` then `tensor.py` imported it

**Review by xurui1995 — APPROVED** [review]

**Review by ZhennanQin — APPROVED** [review]

**Pull request merged**
